src/lib/blkalloc/append_blk_allocator.cpp

```diff
@@ -0,0 +1,180 @@
+}
+
+// alloc a single block;
+BlkAllocStatus AppendBlkAllocator::alloc(BlkId& bid) {
+    if (available_blks() < 1) {
+        COUNTER_INCREMENT(m_metrics, num_alloc_failure, 1);
+        LOGERROR("No space left to serve request nblks: 1, available_blks: {}", available_blks());
+        return BlkAllocStatus::SPACE_FULL;
+    }
+
+    bid.set(m_last_append_offset, 1, m_chunk_id);
+    ++m_last_append_offset;
+    --m_freeable_nblks;
+
+    set_dirty_offset();
+
+    COUNTER_INCREMENT(m_metrics, num_alloc, 1);
+    return BlkAllocStatus::SUCCESS;
+}
+
+//
+// For append blk allocator, the assumption is only one writer will append data on one chunk.
+// If we want to change above design, we can open this api for vector allocation;
+//
+BlkAllocStatus AppendBlkAllocator::alloc(blk_count_t nblks, const blk_alloc_hints& hint,
+                                         std::vector< BlkId >& out_bids) {
+    if (available_blks() < nblks) {
+        COUNTER_INCREMENT(m_metrics, num_alloc_failure, 1);
+        LOGERROR("No space left to serve request nblks: {}, available_blks: {}", nblks, available_blks());
+        return BlkAllocStatus::SPACE_FULL;
+    } else if (nblks > BlkId::max_blks_in_op()) {
+        // consumer(vdev) already handles this case.
+        COUNTER_INCREMENT(m_metrics, num_alloc_failure, 1);
+        LOGERROR("Can't serve request nblks: {} larger than max_blks_in_op: {}", nblks, BlkId::max_blks_in_op());
+        return BlkAllocStatus::FAILED;
+    }
+
+    // Push 1 blk to the vector which has all the requested nblks;
+    out_bids.emplace_back(m_last_append_offset, nblks, m_chunk_id);
+
+    m_last_append_offset += nblks;
+    m_freeable_nblks -= nblks;
+
+    set_dirty_offset();
+
+    COUNTER_INCREMENT(m_metrics, num_alloc, 1);
+
+    return BlkAllocStatus::SUCCESS;
+}
+
+//
+// Alternative design option:
+// Option-1. Everytime offset is updated in a single append_blk_allocator instance, update the dirty buffer into
+// VDevCPContext, which contains a list of (ThreadVector) dirty buffers from all the append_blk_allocator instances
+// (protected by RCU);
+//
+// And cp_flush flush every dirty buffer in that list to disk;
+//
+// Option-2. vdev cp_flush will call each individual chunk's blkallocator and each append_blk_allocator instance flush
+// its dirty buffer on demond;
+//
+// We are using option-2.
+//
+
+//
+// cp_flush is happening in different thread that is running concurrently as AppendBlkAllocator::alloc;
+//
+// 1. if clear_dirty_offset happens before offset is updated by alloc, but before set_dirty_offset, we are flushing
+// offset twice, which is fine.
+// 2. If clear_dirty_offset happens after set_dirty_offset, we are flushing offset ahead of log
+// store, which is also fine during recovery, it is consistent after log store flush completes;
+// 3. Any alloc happened after metablk is persisted will come to next cp;
+//
```

Comment:
I am not sure what you mean by partial update, no persistent is being done here until next_cp.

Reply:
@yamingk I think we can perhaps consider surrounding this with a CPGuard if it is not already done. That way CP will not proceed without exiting this block.

Reply:
@hkadayam I probably didn't get the question, it is during cp_flush state and it is a sync call, why do we want to hold a CPGuard? next_cp won't start until CPManager::on_start_flush completes, right?

Reply:
@yamingk Yes, but while cp is being flushed, the last_append_offset & freeable_nblks could be in a middle of the update and so lets say the io thread updated last_append_offset and not freeable_nblks and in parallel cp is flushing only last_append_offset, but not freeable_nblks there could be inconsistency right?

That is why I was suggesting to put this entire thing in a CPGuard which returns a CPContext pointer and can attach this dirty information into CPContext*. Now it should be safe from concurrent access.

One more benefit is this CPContext* can keep track of all dirty chunks and hence we don't need to rely on HO to persist this or look for active_chunks.

Reply:
@hkadayam Point taken, although I am thinking the other way around: every time a counter/offset are being dirtied (alloc/free), push the dirty buffer to my CPContext. During cp_flush, we just flush whatever was there in my dirty list and switch my next_cp pointing to my next dirty list (RCU). 

My original thought is try not to do the Option-1 since at that time we only have one atomic counter and we don't have race issue with IO thread.  Now that we have freeable_nblks, it turns into a buffer now which gives us no choice but to switch to use Option-1.

Reply:
We don't need to create ThreadVector for this right? CP already provides a safe mechanism for you to protect for race against flushing vs updating (with CPContext).

Now race between concurrent updates of `last_append_offset` and `freeable_nblks` I don't think you need anything other than just atomics like what you have. Nobody is going to read last_append_offset and freeable_nblks together to make any decision yet.

Reply:
We need to discuss whether we can tolerant the race. I am thinking about recovery (log replay) we need to make sure the consistency of these two.